starknet-client/src/client/pathfinder.rs

```diff
@@ -0,0 +1,148 @@
+    Aborted,
+}
+
+impl TransactionStatus {
+    /// Check if the transaction is in a terminal state
+    pub fn is_terminal(&self) -> bool {
+        matches!(
+            self,
+            TransactionStatus::AcceptedOnL1
+                | TransactionStatus::AcceptedOnL2
+                | TransactionStatus::Rejected
+                | TransactionStatus::Reverted
+                | TransactionStatus::Aborted
+        )
+    }
+
+    /// Convert a terminal status into a `Result<(), StarknetClientError>`
+    pub fn into_result(self) -> Result<(), StarknetClientError> {
+        assert!(self.is_terminal());
+        match self {
+            TransactionStatus::AcceptedOnL1 | TransactionStatus::AcceptedOnL2 => Ok(()),
+            TransactionStatus::Rejected => Err(StarknetClientError::TransactionFailure(
+                "transaction rejected".to_string(),
+            )),
+            TransactionStatus::Reverted => Err(StarknetClientError::TransactionFailure(
+                "transaction reverted".to_string(),
+            )),
+            TransactionStatus::Aborted => Err(StarknetClientError::TransactionFailure(
+                "transaction aborted".to_string(),
+            )),
+            _ => unreachable!("non-terminal status found"),
+        }
+    }
+}
+
+/// A raw starknet JSON-RPC request
+#[derive(Clone, Debug, Serialize)]
+pub struct RawJsonRpcRequest {
+    /// The JSON-RPC version
+    pub jsonrpc: String,
+    /// The ID of the request
+    pub id: u64,
+    /// The JSON-RPC method to call
+    pub method: String,
+    /// The parameters to pass to the method
+    pub params: Vec<String>,
+}
+
+/// A raw starknet JSON-RPC response
+#[derive(Clone, Debug, Deserialize)]
+pub struct RawJsonRpcResponse<T>
+where
+    T: for<'de2> Deserialize<'de2>,
+{
+    /// The JSON-RPC version
+    pub jsonrpc: String,
+    /// The ID of the request
+    pub id: u64,
+    /// The result of the request
+    #[serde(bound(deserialize = "for<'de2> T: Deserialize<'de2>"))]
+    pub result: T,
+}
+
+// -------------------------
+// | Client Implementation |
+// -------------------------
+
+impl StarknetClient {
+    /// Get the status of a transaction from the pathfinder API
+    pub async fn get_tx_status(
+        &self,
+        tx_hash: FieldElement,
+    ) -> Result<TransactionStatus, StarknetClientError> {
+        let tx_hash_str = format!("0x{tx_hash:x}");
+        self.raw_json_rpc_request(GET_TRANSACTION_STATUS_METHOD.to_string(), vec![tx_hash_str])
+            .await
+    }
+
+    /// Make a raw JSON-RPC request to the RPC node
+    async fn raw_json_rpc_request<T: for<'de> Deserialize<'de>>(
+        &self,
+        method: String,
+        params: Vec<String>,
+    ) -> Result<T, StarknetClientError> {
+        let res = self
+            .http_client
+            .post(self.config.starknet_json_rpc_addr.clone())
+            .json(&RawJsonRpcRequest {
+                jsonrpc: JSON_RPC_VERSION.to_string(),
+                id: 1,
```

Comment:
should we make the `id` an incrementing counter?

Reply:
These are synchronous calls, I don't see what we gain by maintaining more state?